src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1140 @@
+/// The credRandom is used as a secret to HMAC those salts.
+/// The last step is to re-encrypt the outputs.
+fn encrypt_hmac_secret_output(
+    shared_secret: &[u8; 32],
+    salt_enc: &[u8],
+    cred_random: &[u8],
+) -> Result<Vec<u8>, Ctap2StatusCode> {
+    if salt_enc.len() != 32 && salt_enc.len() != 64 {
+        return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+    }
+    if cred_random.len() != 32 {
+        // We are strict here. We need at least 32 byte, but expect exactly 32.
+        return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+    }
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(shared_secret);
+    let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+    // The specification specifically asks for a zero IV.
+    let iv = [0u8; 16];
+
+    let mut cred_random_secret = [0u8; 32];
+    cred_random_secret.clone_from_slice(cred_random);
+
+    // Initialization of 4 blocks in any case makes this function more readable.
+    let mut blocks = [[0u8; 16]; 4];
+    // With the if clause restriction above, block_len can only be 2 or 4.
+    let block_len = salt_enc.len() / 16;
+    for i in 0..block_len {
+        blocks[i].copy_from_slice(&salt_enc[16 * i..16 * (i + 1)]);
+    }
+    cbc_decrypt(&aes_dec_key, iv, &mut blocks[..block_len]);
+
+    let mut decrypted_salt1 = [0u8; 32];
+    decrypted_salt1[..16].clone_from_slice(&blocks[0]);
+    let output1 = hmac_256::<Sha256>(&cred_random_secret, &decrypted_salt1[..]);
+    decrypted_salt1[16..].clone_from_slice(&blocks[1]);
+    for i in 0..2 {
+        blocks[i].copy_from_slice(&output1[16 * i..16 * (i + 1)]);
+    }
+
+    if block_len == 4 {
+        let mut decrypted_salt2 = [0u8; 32];
+        decrypted_salt2[..16].clone_from_slice(&blocks[2]);
+        decrypted_salt2[16..].clone_from_slice(&blocks[3]);
+        let output2 = hmac_256::<Sha256>(&cred_random_secret, &decrypted_salt2[..]);
+        for i in 0..2 {
+            blocks[i + 2].copy_from_slice(&output2[16 * i..16 * (i + 1)]);
+        }
+    }
+
+    cbc_encrypt(&aes_enc_key, iv, &mut blocks[..block_len]);
+    let mut encrypted_output = Vec::with_capacity(salt_enc.len());
+    for b in &blocks[..block_len] {
+        encrypted_output.extend(b);
+    }
+    Ok(encrypted_output)
+}
+
+fn check_and_store_new_pin(
+    persistent_store: &mut PersistentStore,
+    aes_dec_key: &crypto::aes256::DecryptionKey,
+    new_pin_enc: Vec<u8>,
+) -> bool {
+    if new_pin_enc.len() != PIN_PADDED_LENGTH {
+        return false;
+    }
+    let iv = [0u8; 16];
+    // Assuming PIN_PADDED_LENGTH % block_size == 0 here.
+    let mut blocks = [[0u8; 16]; PIN_PADDED_LENGTH / 16];
+    for i in 0..PIN_PADDED_LENGTH / 16 {
+        blocks[i].copy_from_slice(&new_pin_enc[i * 16..(i + 1) * 16]);
+    }
+    cbc_decrypt(aes_dec_key, iv, &mut blocks);
+    let mut pin = vec![];
+    'pin_block_loop: for block in blocks.iter().take(PIN_PADDED_LENGTH / 16) {
```

Comment:
This loop can be replaced by `let pin: Vec<u8> = blocks.iter().take(n).flatten().collect();`

Reply:
I missed that part. You can chain with `.take_while(|&c| c != 0)` before the `.collect()` and this should be good to go.

Reply:
Okay, I added a comment that we use the implementation that is is compatible with 2.0. The behavior is different for padded PINs like

> ['1', '2', '3', '4', '\0', '6', '\0', ...]

In 2.0, the PIN is read as "1234". In 2.1 instead, two more bytes are included. IMO we should never receive padded PINs like that anyway.

Reply:
That's IMO a good reason to separate the logic into a `cbc_decrypt_pin` function. Then it's easy to make unit tests like `assert_eq!(cbc_decrypt_pin(encrypt(['1', '2', '3', '4', 0, '6', 0, ...])), ['1', '2', '3', '4'])`